flang/lib/Lower/OpenMP.cpp

```diff
@@ -87,19 +88,35 @@ genOMP(Fortran::lower::AbstractConverter &absConv,
     auto &firOpBuilder = absConv.getFirOpBuilder();
     auto currentLocation = absConv.getCurrentLocation();
     auto insertPt = firOpBuilder.saveInsertionPoint();
+
+    // Clauses.
+    mlir::Value numThreads;
+
+    const auto &parallelOpClauseList =
+        std::get<Fortran::parser::OmpClauseList>(blockDirective.t);
+    for (const auto &clause : parallelOpClauseList.v) {
+      if (const auto &numThreadsClause =
+              std::get_if<Fortran::parser::OmpClause::NumThreads>(&clause.u)) {
+        numThreads = firOpBuilder.createIntegerConstant(
+            currentLocation, firOpBuilder.getIntegerType(32),
+            Fortran::evaluate::ToInt64(
+                *Fortran::semantics::GetExpr(numThreadsClause->v))
+                .value());
+      }
+    }
     llvm::ArrayRef<mlir::Type> argTy;
-    mlir::ValueRange range;
-    llvm::SmallVector<int32_t, 6> operandSegmentSizes(6 /*Size=*/,
-                                                      0 /*Value=*/);
-    // create and insert the operation.
+    // TODO: Add support for remaining clauses.
+    mlir::ValueRange range(numThreads);
+    // Create and insert the operation.
     auto parallelOp = firOpBuilder.create<mlir::omp::ParallelOp>(
         currentLocation, argTy, range);
-    parallelOp.setAttr(mlir::omp::ParallelOp::getOperandSegmentSizeAttr(),
-                       firOpBuilder.getI32VectorAttr(operandSegmentSizes));
-    parallelOp.getRegion().push_back(new Block{});
+    parallelOp.setAttr(
+        mlir::omp::ParallelOp::getOperandSegmentSizeAttr(),
+        firOpBuilder.getI32VectorAttr({0, (numThreads ? 1 : 0), 0, 0, 0, 0}));
```

Comment:
What are all these magic 0 constants? Could a comment with a reference or description be given?

Reply:
This and Range creation when NUM_THREAD clause is not present is still to be addressed.